spring-integration-ftp/src/main/java/org/springframework/integration/ftp/gateway/FtpOutboundGateway.java

```diff
@@ -151,9 +185,75 @@ protected long getModified(FTPFile file) {
 	}
 
 	@Override
+	protected void doInit() {
+		super.doInit();
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+	}
+
+	@Override
 	protected FTPFile enhanceNameWithSubDirectory(FTPFile file, String directory) {
 		file.setName(directory + file.getName());
 		return file;
 	}
 
+	@Override
+	protected List<?> ls(Message<?> message, Session<FTPFile> session, String dir) throws IOException {
+		return doInWorkingDirectory(message, session,
+				() -> super.ls(message, session, dir));
+	}
+
+	@Override
+	protected List<String> nlst(Message<?> message, Session<FTPFile> session, String dir) throws IOException {
+		return doInWorkingDirectory(message, session,
+				() -> super.nlst(message, session, dir));
+	}
+
+	@Override
+	protected File get(Message<?> message, Session<FTPFile> session, String remoteDir, String remoteFilePath,
+			String remoteFilename, FTPFile fileInfoParam) throws IOException {
+		return doInWorkingDirectory(message, session,
+				() -> super.get(message, session, remoteDir, remoteFilePath, remoteFilename, fileInfoParam));
+	}
+
+	@Override
+	protected List<File> mGet(Message<?> message, Session<FTPFile> session, String remoteDirectory,
+			String remoteFilename) throws IOException {
+		return doInWorkingDirectory(message, session,
+				() -> super.mGet(message, session, remoteDirectory, remoteFilename));
+	}
+
+	@Override
+	protected boolean rm(Message<?> message, Session<FTPFile> session, String remoteFilePath) throws IOException {
+		return doInWorkingDirectory(message, session,
+				() -> super.rm(message, session, remoteFilePath));
+	}
+
+	private <V> V doInWorkingDirectory(Message<?> message, Session<FTPFile> session, Callable<V> task)
+			throws IOException {
+		Expression workingDirExpression = this.workingDirExpression;
+		FTPClient ftpClient = (FTPClient) session.getClientInstance();
+		String currentWorkingDirectory = null;
+		try {
+			if (workingDirExpression != null) {
+				currentWorkingDirectory = ftpClient.printWorkingDirectory();
+				String workingDir = workingDirExpression.getValue(this.evaluationContext, message, String.class);
+				ftpClient.changeWorkingDirectory(workingDir);
```

Comment:
Well, we can't assume that user uses some other working dir by default and decided for some messages to switch to his/her root. And, of course, then come back after the operation. Remember `AbstractFtpSessionFactory.postProcessClientBeforeConnect/postProcessClientAfterConnect` ?

> The working directory can be set via the FTPClient.changeWorkingDirectory() function when you extend the DefaultFtpSessionFactory and implement postProcessClientAfterConnect() callback.